Adatb1/2zh/feladat9_plsql_megoldasok.txt

```diff
@@ -0,0 +1,200 @@
+  TYPE Map_Dnev_To_Fiz IS
+    TABLE OF dolg2.fizetes%TYPE
+    INDEX BY dolg2.dnev%TYPE;
+  mymap Map_Dnev_To_Fiz;
+  isOddRow Boolean := true;
+begin
+  for currentRow in (
+      select dnev, fizetes
+      from dolg2
+      order by dnev)
+  loop
+    if isOddRow then
+      mymap(currentRow.dnev) := currentRow.fizetes;
+    end if;
+    isOddRow := not isOddRow;
+  end loop;
+
+  dbms_output.put_line('dnev: '||mymap.prior(mymap.last)||', fizetes: '||mymap(mymap.prior(mymap.last)));
+end;
+/
+set serveroutput on
+call proc9();
+
+
+/* (paraméteres kurzor)
+Írjunk meg egy plsql programot (név nélküli blokkot), amelyik kiírja azon dolgozók
+nevét és belépési dátumát, akik a felhasználó által megadott osztályon dolgoznak.
+A felhasználó az osztály nevének első betűjét adja meg (ACCEPT-tel kérjük be).
+A program írja ki az osztály nevét is. Ha nincs megfelelő osztály, akkor azt írja ki.
+*/
+ACCEPT oazon_begin CHAR FORMAT '9' PROMPT 'Adja meg az osztály azonosító első betűjét:';
+declare
+  oazon_in oszt2.oazon%type := null;
+  osztNev  oszt2.onev%type;
+begin
+  select oazon into oazon_in from oszt2
+  where oazon LIKE &oazon_begin||'%';
+
+  select onev into osztNev from oszt2 where oazon = oazon_in;
+  dbms_output.put_line('Talált osztály: '||osztNev||'.');
+
+  for curr in (select dnev, belepes from dolg2 where oazon = oazon_in) loop
+    dbms_output.put_line('Dnev: '||curr.dnev||', belepes: '||curr.belepes);
+  end loop;
+exception
+  when NO_DATA_FOUND then
+    dbms_output.put_line('Nincs megfelelő osztály.');
+end;
+/
+
+
+/* (módosítás)
+Írjunk meg egy procedúrát, amelyik megnöveli azoknak a dolgozóknak a fizetését 1-el,
+akiknek a fizetési kategóriája ugyanaz, mint a procedúra paramétere.
+A procedúra a módosítás után írja ki a módosított (új) fizetések átlagát két tizedesjegyre kerekítve.
+*/
+CREATE OR REPLACE
+PROCEDURE kat_novel(p_kat NUMBER) IS
+  avg_fiz real;
+begin
+  update dolg2
+  set fizetes = fizetes + 1
+  where dkod in (
+    select dkod
+    from dolg2 d
+    join fizkat2 fk
+      on d.fizetes between fk.also and fk.felso
+    where kategoria = p_kat
+  );
+
+  select avg(fizetes) into avg_fiz
+  from dolg2
+  where dkod in (
+    select dkod
+    from dolg2 d
+    join fizkat2 fk
+      on d.fizetes between fk.also and fk.felso
+    where kategoria = p_kat
+  );
+
+  dbms_output.put_line('A módosított dolgozók átlagfizetése: '|| avg_fiz);
+  ROLLBACK;
+end;
+/
+set serveroutput on
+call kat_novel(2);
+
+
+/* (módosítás kurzorral)
+Írjunk meg egy procedúrát, amelyik módosítja a paraméterében megadott osztályon a fizetéseket, és
+kiírja a dolgozó nevét és új fizetését.  A módosítás mindenki fizetéséhez adjon hozzá n*10 ezret,
+ahol n a dolgozó nevében levő magánhangzók száma (a, e, i, o, u).
+*/
+CREATE OR REPLACE
+PROCEDURE fiz_mod(p_oazon INTEGER) IS
+  cursor dolgCurs is
+    select dnev, fizetes
+    from dolg2
+    where oazon = p_oazon
+  for update;
+
+  currDnev dolg2.dnev%type;
+  currFiz  dolg2.fizetes%type;
```

Comment:
Why use two a variable??
Use one: Example:
rec dolgCurs%ROWTYPE;
fetch dolgCurs into rec;
and use:
rec.dnev or rec.fizetes